File: tests/lib/env.ts
import { readFileSync } from "fs";
import { resolve } from "path";
import dotenv from "dotenv";

let loaded = false;

export function loadEnv(): void {
  if (loaded) return;
  loaded = true;

  // Load backend/.env for secrets
  const envPath = resolve(import.meta.dirname, "../../backend/.env");
  try {
    const content = readFileSync(envPath, "utf-8");
    const parsed = dotenv.parse(content);
    for (const [key, value] of Object.entries(parsed)) {
      if (!process.env[key]) {
        process.env[key] = value;
      }
    }
  } catch {
    // .env file is optional if vars are set in environment
  }
}

export function requireEnv(key: string): string {
  loadEnv();
  const value = process.env[key];
  if (!value) {
    throw new Error(`Missing required environment variable: ${key}. Set it in env or backend/.env`);
  }
  return value;
}

File: test/e2e/lib/env.py
from __future__ import annotations

import os
from pathlib import Path

from dotenv import dotenv_values

_loaded = False

SERVER_DIR = Path(__file__).resolve().parents[3] / "server"


def load_env() -> None:
    global _loaded
    if _loaded:
        return
    _loaded = True

    env_path = SERVER_DIR / ".env"
    if env_path.exists():
        for key, value in dotenv_values(env_path).items():
            if value is not None:
                os.environ.setdefault(key, value)


def require_env(key: str) -> str:
    load_env()
    value = os.environ.get(key)
    if not value:
        raise RuntimeError(f"Missing required environment variable: {key}. Set it in env or server/.env")
    return value
